executor/analyze.go

```diff
@@ -506,12 +525,453 @@ type AnalyzeFastExec struct {
+		if *err != nil {
+			return
+		}
+		ctx := context.Background()
+		resp, *err = client.SendRequest(ctx, rpcCtx.Addr, req, tikv.ReadTimeoutMedium)
+		if *err != nil {
+			return
+		}
+		// TODO: check the region not_found
+		if ctx.Err() != nil {
+			fmt.Println(ctx.Err().Error())
+			*needRebuild = true
+			return
+		}
+		for _, prop := range resp.DebugGetRegionProperties.Props {
+			if prop.Name == "num_rows" {
+				var cnt uint64
+				cnt, *err = strconv.ParseUint(prop.Value, 10, 64)
+				if *err != nil {
+					return
+				}
+				newCount := atomic.AddUint64(&e.rowCount, cnt)
+				task := &AnalyzeFastTask{
+					Location:  loc,
+					LRowCount: newCount - cnt,
+					RRowCount: newCount,
+				}
+				*sampTasks = append(*sampTasks, task)
+			}
+		}
+	}
+}
+
+// buildSampTask return tow variable, the first bool is whether the task meeting region error
+// and need to rebuild.
+func (e *AnalyzeFastExec) buildSampTask() (bool, error) {
+	// Do get regions row count.
+	bo := tikv.NewBackoffer(context.Background(), 500)
+	atomic.StoreUint64(&e.rowCount, 0)
+	needRebuildForRoutine := make([]bool, e.concurrency)
+	errs := make([]error, e.concurrency)
+	sampTasksForRoutine := make([][]*AnalyzeFastTask, e.concurrency)
+	e.sampLocs = make(chan *tikv.KeyLocation, e.concurrency)
+	e.wg.Add(e.concurrency)
+	for i := 0; i < e.concurrency; i++ {
+		go e.getSampRegionsRowCount(bo, &needRebuildForRoutine[i], &errs[i], &sampTasksForRoutine[i])
+	}
+
+	store, _ := e.ctx.GetStore().(tikv.Storage)
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.PhysicalTableID)
+	var loc *tikv.KeyLocation
+	var err error
+	// extract all regions contain the table
+	e.scanTasks = e.scanTasks[:0]
+	e.sampTasks = e.sampTasks[:0]
+	// If length of KeyLocation.StartKey is 0, then the start key is infinitesimal.
+	// If length of KeyLocation.EndKey is 0, then the end kety is infinity.
+	for loc, err = e.cache.LocateKey(bo, startKey); bytes.Compare(loc.StartKey, endKey) <= 0 && err == nil; loc, err = e.cache.LocateKey(bo, loc.EndKey) {
+		if bytes.Compare(endKey, loc.EndKey) < 0 || bytes.Compare(loc.StartKey, startKey) < 0 || len(loc.StartKey) == 0 || len(loc.EndKey) == 0 {
+			e.scanTasks = append(e.scanTasks, loc)
+			if len(loc.StartKey) == 0 {
+				loc.StartKey = startKey
+			}
+			if len(loc.EndKey) == 0 {
+				loc.EndKey = endKey
+				break
+			}
+		} else {
+			e.sampLocs <- loc
+		}
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	close(e.sampLocs)
+	e.wg.Wait()
+	for i := 0; i < e.concurrency; i++ {
+		if errs[i] != nil {
+			return false, errors.Trace(errs[i])
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		if needRebuildForRoutine[i] == true {
+			return true, nil
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		e.sampTasks = append(e.sampTasks, sampTasksForRoutine[i]...)
+	}
+
+	return false, nil
+}
+
+func decodeValues(sValue []byte, sKey kv.Key) (values []types.Datum, err error) {
```

Comment:
This function did not take the default value of the column into consideration, you could take the implementation of point get as an example.